arch/x86/core/userspace.S

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <arch/x86/asm.h>
+#include <offsets_short.h>
+
+GTEXT(_x86_syscall_invoke)
+GTEXT(_x86_syscall_entry_stub)
+GTEXT(_x86_syscall_entry)
+GTEXT(_x86_userspace_enter)
+
+/* Landing site for syscall SW IRQ. Marshal arguments and call C function for
+ * further processing.
+ */
+SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
+	push	%esi		/* arg5 */
+	push	%edi		/* arg4 */
+	push	%ebx		/* arg3 */
+#ifndef CONFIG_X86_IAMCU
+	push	%ecx		/* arg2 */
+	push	%edx		/* arg1	*/
+	push	%eax		/* Call id */
+#endif
+
+	call	_x86_syscall_entry
+
+	/* EAX now contains return value. Pop or xor everything else to prevent
+	 * information leak from kernel mode.
+	 */
+#ifndef CONFIG_X86_IAMCU
+	add	$4, %esp	/* old EAX value */
+	pop	%edx
+	pop	%ecx
+#else
+	xor	%edx, %edx
+	xor	%ecx, %ecx
+#endif
+	pop	%ebx
+	pop	%edi
+	pop	%esi
+
+	iret
+
+
+/* FUNC_NORETURN void _x86_userspace_enter(_thread_entry_t user_entry,
+ *					   void *p1, void *p2, void *p3,
+ *					   u32_t stack_end,
+ *					   u32_t stack_start)
+ *
+ * A one-way trip to userspace.
+ */
+SECTION_FUNC(TEXT, _x86_userspace_enter)
+	add	$0x4, %esp /* Discard return address on stack */
+
+	/* Fetch parameters on the stack */
+#ifndef CONFIG_X86_IAMCU
+	pop	%eax	/* user_entry */
+	pop	%edx	/* p1 */
+	pop	%ecx	/* p2 */
+#endif
+	pop	%esi	/* p3 */
+	pop	%ebx	/* stack_end (high address) */
+	pop	%edi	/* stack_start (low address) */
+
+	/* Move to the kernel stack for this thread, so we can erase the
+	 * user stack. The kernel stack is the page immediately before
+	 * the user stack.
+	 *
+	 * For security reasons, we must erase the entire user stack.
+	 * We don't know what previous contexts it was used and do not
+	 * want to leak any information.
```

Comment:
Not loving this.  Obviously yes: in the abstract a thread stack might have anything in it and we need to start from a known state.  But the overwhelming majority of these stacks are going to be static things whose stacks where started from kernel mode at system startup and which haven't been presented to anything but OS code yet.  There ought to be a way to capture that info somewhere to avoid the double-memset.

Maybe something that would be better captured by a jira to track the needed optimization.

Also: why is this being done in assembly?  Can't a precondition for calling this simply be "the stack must be zero" and so whatever kernel code is responsible for doing that does its own memset() in C-land?  It's not like it's complicated code, but as we grow this is going to have to be rewritten (in assembly!) for every architecture.

Reply:
The stack we are zeroing is for the currently running thread, if you memset while on that stack you will explode. at minimum, you would have to switch to the kernel stack first, then call memset. but perhaps I can make a call to memset here. let me try this.

> But the overwhelming majority of these stacks are going to be static things whose stacks where started from kernel mode at system startup and which haven't been presented to anything but OS code yet

I was thinking we would implement everything with extreme paranoia for the first pass, and then later on either introduce kconfig options or runtime flags to relax things.

Reply:
the stack has to be zeroed here, as you cannot zero out a stack you are currently on,
I played around with calling memset from here, but it's soooo many more instructions than doing a 'rep stos'. if other arches want to call memset they can, as long as they switch to the kernel stack first. i'd like to leave this alone on x86.